### PR TITLE
Callable mag, dex, dB units, and parsing of such units.

### PR DESCRIPTION
**Pull request by mhvk**
EDIT -- now rewritten such that one just does `u.mag(u.ct/u.s)`, where `u.mag` is a standard unit that has a `__call__` method.

ORIGINAL TEXT:

Just as a proof of concept, this allows one to do, e.g.,
```
import astropy.units as u
from astropy.units.function import mag, dex, dB
mag(u.ct/u.s)
# MagUnit('ct / s')
```
To be useful, it would imply using the function unit `mag` as the default instance for `u.mag`. This would require more work.

### CODE REVIEW

**Comment by embray** [comment]
I haven't looked at the code, but I like the concept!

**Comment by mhvk** [comment]
Note to self: if we do this, `u.mag` should almost certainly refer to `units.function.mag` rather than `units.astrophys.mag` -- the latter would just be used internally in the function units (or in composite units).

**Comment by mhvk** [comment]
Set the 1.10 milestone, since I think with this merged, the preferred API for function units changes: one would do `u.dex(u.cm/u.s**2)` instead of `u.DexUnit(u.cm/u.s**2)`. And if this is still done in the same milestone, we don't have to make any explicit notices about preferred API.

**Comment by mhvk** [comment]
@mdboom - here's an update on the callable units. It now makes `mag`, `dex`, and `dB` be subclass instances of their normal classes which have an additional `__call__` method. Can you have a look that you think the implementation is OK? In principle, one could also add a `__call__` method to `UnitBase` (which normally would raise a `TypeError`), but I thought it was better to keep the changes completely separate. It does mean that `mag`, `dex` and `dB` are now defined in `function.units` instead of in `astrophys`, but that seems, if anything, a more logical place (and in practice one would just get them from the top level anyway).

I'll try to update the documentation later today.

File: astropy/units/function/mixin.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    def __call__(self, unit=None):
+        if self._function_unit_class is None:
+            raise TypeError("Cannot call unit since corresponding "
+                            "function unit class is not defined.")
+        return self._function_unit_class(physical_unit=unit,
+                                         function_unit=self)
+
+
+class IrreducibleUnit(FunctionMixin, core.IrreducibleUnit):
```
Comment (mdboom) [inline astropy/units/function/mixin.py]: I think reusing the same name here is a bit confusing (when used in the `function/units.py` file below).  Maybe call it `IrreducibleFunctionUnit`. Same for `FunctionUnit` below...

**Comment by mdboom** [comment]
Looks good other than my one minor comment.

**Comment by mhvk** [comment]
@mdboom - OK, changed those names. I now also updated the documentation, including giving a little teaser in the "getting started" section of units.

**Comment by mhvk** [comment]
@mdboom - I think this is now all done  (I now use the callable units in the parser and resolved a small issue found while testing with `__numpy_ufunc__`); if you want to check anything, perhaps the revised documentation?

**Comment by mhvk** [comment]
@mdboom - would you want/have a chance to have a last look at this PR? (And the follow-up in #3851...)

**Pull request merged**